### PR TITLE
Feature/houses

### PR DESCRIPTION
**Pull request by nsixcancode**
Phase 1 of las houses
- Create functionality that adds up the time each dad spent drawing
- Create plots of that data and put it on a prototype leader board page
  - This page will ONLY be a prototype for the next phase
  - This will take all users and plot them all essentially on the same house

You can see this graph on the /houses endpoint

Stylistic input is definitely needed.

### CODE REVIEW

File: db/migrate/20200823015502_create_house_participations.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+# Migration to create house participations
+class CreateHouseParticipations < ActiveRecord::Migration[6.0]
+  def change
+    create_table :house_participations do |t|
+      t.integer :user_id, null: false
+      t.integer :house_id, null: false
+      t.date :join_date, null: false
+      t.integer :score, null: false
```
Comment (the1banana) [inline db/migrate/20200823015502_create_house_participations.rb:10]: This has me thinking, we do a lot of things with regards to submissions being converted to points at the get go.
But it's entirely possible for people to submit multiple short submissions.
Let's say they spent 20 minutes on three submissions. Each of these would contribute 0 points, but he's done an hour of work and should be entitled to an hour!
Could we either opt for fractional amounts or have the system track these pieces properly? I don't want to force people to compile things together for the sake of houses alone.

File: devScripts/fillDatabase.sql [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+        insert into house_participations (user_id, house_id, join_date)
+        values ((select id from users where name = cast (counter as varchar)), floor(random()*3+4), CURRENT_TIMESTAMP);
+
+        FOR user_submissions IN 1..floor(random()*30) LOOP
+            insert into submissions (user_id, time, created_at, updated_at)
+            values ((select id from users where name = cast (counter as varchar)), floor(random()*100), CURRENT_TIMESTAMP, CURRENT_TIMESTAMP);
+        END LOOP;
+    END LOOP;
+END; $$
```
Comment (the1banana) [inline devScripts/fillDatabase.sql:25]: this is very much appreciated, but if you could make this a rake task instead of a .sql file, that would be great.

File: lib/tasks/dad_tasks.rake [hunk 3]
```diff
@@ -1,437 +1,447 @@
-# frozen_string_literal: true
```
Comment (the1banana) [inline lib/tasks/dad_tasks.rake:1]: diff isn't helpful here... could you summarize what was changed?

**Review by the1banana — CHANGES_REQUESTED** [review]
some comments, great work!

File: devScripts/fillDatabase.sql (hunk 2 above)
Comment (nsixcancode) [reply devScripts/fillDatabase.sql:25]: I actually didn't mean to commit that. I should get rid of it.

File: lib/tasks/dad_tasks.rake (hunk 3 above)
Comment (nsixcancode) [reply lib/tasks/dad_tasks.rake:1]:
```
    # STEP 6: Create new Houses if it is the start of the month
    if Time.now.utc.to_date == Time.now.utc.at_beginning_of_month.to_date
      (1..3).each do |i|
        House.create(
          house_name: "House #{i}",
          house_start: Time.now.utc.at_beginning_of_month.to_date,
        )
      end
    end
```

This was all I added. It just checks to see if it's the first of the month, and if it is, it makes 3 new houses.

File: db/migrate/20200823015502_create_house_participations.rb (hunk 1 above)
Comment (nsixcancode) [reply db/migrate/20200823015502_create_house_participations.rb:10]: I asked las houses about this and they were pretty set on 30mins = 1pt. No fractions, no nothing. The way it currently works in las is people are allowed to round up/down if its a few minutes under/over. Combining submissions is not uncommon in houses. 

I suppose I could add up their total amount of time for the month and then integer divide by 30.

I would like keep our implementation as close to las houses as possible. I can ask them again what they think.

File: app/controllers/houses_controller.rb [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    params.require(:house).permit(:house_name)
+  end
+
+  # @function sane_date
+  # @abstract this is because the date selectors for rails suck.
+  def sane_date
+    return params[:date] unless params[:date].is_a?(ActionController::Parameters)
+
+    params[:date] = params[:date][:year] + '-' + params[:date][:month]
+  end
```
[the remark this answers is not in the document]
Comment (nsixcancode) [reply app/controllers/houses_controller.rb:66]: Making a note that this has been addressed in my next commit. (ignore this)

File: app/controllers/houses_controller.rb [hunk 5]
```diff
@@ -0,0 +1,67 @@
+# frozen_string_literal: true
+
+# Controller for houses
+class HousesController < ApplicationController
+  include HousesHelper
+  before_action :ensure_moderator, only: %i[edit update]
```
[the remark this answers is not in the document]
Comment (nsixcancode) [reply app/controllers/houses_controller.rb:6]: Making a note that this has been addressed in my next commit. (ignore this)

File: app/models/house.rb [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+  # @function place
+  # @return the ranked place compared ot rival houses based on score ("first", "second", or "third")
+  # @note teams can tie for first and second.
+  def place
+    return 'first' if first?
+
+    return 'second' if second?
+
+    'third'
+  end
```
[the remark this answers is not in the document]
Comment (nsixcancode) [reply app/models/house.rb:48]: I generalized it like so:
```
def place
  (1 + House
    .where('house_start = ?', house_start)
    .sort_by { |h| -h.total }
    .find_index { |h| h.id = id }).ordinalize
end
```

What I don't like about this is there is no way to "tie" if two teams have the same score. also there is no way know how sort_by will arrange the teams if they are equal.

Statistically, it is unlikely that two teams will tie, so it shouldn't matter. But edge cases...

File: app/views/houses/index.html.slim [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+      a.btn.btn-primary href="#{houses_path(:date => @date.next_month.strftime('%Y-%m'))}"
+        | Next Month
+
+  div.house-neighborhood
+    - if @houses.any?
+      - @houses.each_with_index do |h, i|
+        div class="card house-card"
+          .card-header.house-card-header
+            b style="flex: 1" #{h.house_name}  
+            | #{"🥇" if h.place == 'first'} #{"🥈" if h.place == 'second'} #{"🥉" if h.place == 'third'}
```
[the remark this answers is not in the document]
Comment (nsixcancode) [reply app/views/houses/index.html.slim:25]: Emoji's are part of unicode and read just like normal text. Most modern operating systems support them either natively or with a package.
https://en.wikipedia.org/wiki/Emoji#Modern-day_emojis_(2010%E2%80%93present)

I would have used an icon, but font awesome doesn't have a first second and third trophy/medal.

I'll take it out if you don't like it. I figured since it was just HTML it wouldn't be a big deal and this seemed like the simplest solution. (I kinda like it to be honest 😂)

File: app/models/house_participation.rb [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    self.score -= points
+    save
+  end
+
+  # @function update_points
+  # @abstract removes points and then adds a different
+  # amount of points (when updating submission time)
+  # @does not update if the submission is older than a month.
+  def update_points(old_points, new_points, date)
+    return unless date.to_date >= join_date.at_beginning_of_month.to_date
```
[the remark this answers is not in the document]
Comment (nsixcancode) [reply app/models/house_participation.rb:27]: You are right. I used join_date instead of current date. My bad.

File: app/controllers/submissions_controller.rb [hunk 9]
```diff
@@ -128,6 +136,18 @@ def update
...
     end
 
+    # Modify their points if they change their time spent on a submission
+    if used_params.has_key? :time
+      current_user.house_participations
+          .where('join_date >=  ?', Time.now.utc.at_beginning_of_month.to_date)
+          &.first
+          &.update_points(
+            @submission.time.to_i / 30,
+            used_params[:time].to_i / 30,
```
[the remark this answers is not in the document]
Comment (nsixcancode) [reply app/controllers/submissions_controller.rb:146]: We're switching to raw time so this doesn't matter anymore.

File: db/migrate/20200823015502_create_house_participations.rb (hunk 1 above)
Comment (nsixcancode) [reply db/migrate/20200823015502_create_house_participations.rb:10]: We're using raw time. No point conversion.

File: app/assets/javascripts/houses.coffee [hunk 10]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (the1banana) [inline app/assets/javascripts/houses.coffee:1]: this doesn't do anything, and lets try not to use coffeescript anymore

File: app/controllers/houses_controller.rb [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+
+  # @function sane_date
+  # @abstract this handles the date_selector returning a hash rather than a string
+  # If the user uses the dropdown, it converts the date hash to a string.
+  # If the user uses the buttons or doesn't pass anything, it just uses the string or nothing.
+  def sane_date
+    return params[:date] unless params[:date].is_a?(ActionController::Parameters)
+
+    params[:date] = params[:date][:year] + '-' + params[:date][:month]
+  end
```
Comment (the1banana) [inline app/controllers/houses_controller.rb:76]: hella weird, but maybe this is a weird ruby-ism
if it works it works though
we can talk about refactoring this later

File: app/models/house.rb [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+  # @function add_user
+  # @abstract adds the current user
+  def add_user(uid)
+    return can_join?(uid) unless can_join?(uid).blank?
+
+    score = User.find(uid).submissions
+                .where('created_at >= ?', search_date).sum('submissions.time').to_i
+    HouseParticipation
+      .create(user_id: uid, house_id: id, join_date: Time.now.utc.to_date, score: score)
+    ''
```
Comment (the1banana) [inline app/models/house.rb:20]: what is this single quote doing here?

File: app/models/house_participation.rb [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+  # @note not to be used when a user is banned
+  def remove_points(points)
+    self.score -= points
+    save
+  end
+
+  # @function update_points
+  # @abstract removes points and then adds a different
+  # amount of points (when updating submission time)
+  # does not update if the submizssion is older than a month.
```
Comment (the1banana) [inline app/models/house_participation.rb:26]: "submizssion"

**Review by the1banana — APPROVED** [review]
Extremely close. Just make those really minor corrections and I'll merge it in ASAP.

File: app/models/house.rb (hunk 12 above)
Comment (nsixcancode) [reply app/models/house.rb:20]: This is probably my inexperience with Ruby. When there is an error in joining houses, the function returns a string with context for why the user couldn't join a house. When there is no error, the function will ideally return nil/nothing, but because ruby returns the last line of code, it would return the "HouseParticipation" instance that was created. So instead I just had it return an empty string and had it check to see if the string was empty or not to indicate whether or not there was an error. Probably not the best way of doing that I can I can fix that.

File: app/models/house_participation.rb (hunk 13 above)
Comment (nsixcancode) [reply app/models/house_participation.rb:26]: me fail english? that's unpossible.
